File: Programa/logica/tablero.py
from Casilla import TipoCasilla
from Casilla import Casilla
from Jugador import Color

class Tablero:
	jugadores = []
	tablero = []
	pasilloAmarillo = []
	pasilloAzul = []
	pasilloRojo = []
	pasilloVerde = []

	def __init__(self):
		self.crearTablero()
		self.crearPasillos()
	
	#agrega un jugador al tablero
	def agregarJugador(self,jugador):
		if(len(self.jugadores)<4):
			self.jugadores.append(jugador)

	#crea el tablero de parchis
	def crearTablero(self):
		casillasSeguras = [12,17,29,34,46,51,63,68]
		for i in range(1,69):
			if(i== 5):
				self.tablero.append(Casilla(TipoCasilla.inicio,Color.Amarillo,i))
			elif(i== 22):
				self.tablero.append(Casilla(TipoCasilla.inicio,Color.Azul,i))
			elif(i== 39):
				self.tablero.append(Casilla(TipoCasilla.inicio,Color.Rojo,i))
			elif(i== 56):
				self.tablero.append(Casilla(TipoCasilla.inicio,Color.Verde,i))
			elif(i in casillasSeguras):
				self.tablero.append(Casilla(TipoCasilla.segura,None,i))
			else:
				self.tablero.append(Casilla(TipoCasilla.normal,None,i))
	
	#Devuelve la casilla en la posicion indicada, de forma circular (1 es el primer indice)
	def obtenerCasilla(self,posicion):
		return self.tablero[(posicion-1)%len(self.tablero)]

	#Devuelve el pasillo de un color
	def obtenerListaPasillo(self,color):
		if(color == Color.Amarillo):
			return self.pasilloAmarillo
		elif(color == Color.Azul):
			return self.pasilloAzul
		elif(color == Color.Rojo):
			return self.pasilloRojo
		elif(color == Color.Verde):
			return self.pasilloVerde
		else:
			return None

	#Devuelve la lista de jugadores
	def obtenerPasillo(self,jugador,posicion):
		lista = self.obtenerListaPasillo(jugador.color)
		if(posicion < len(lista)):
			return lista[posicion]

	#Crea los pasillos de cada color
	def crearPasillos(self):
		for i in range(8):
			if(i<7):
				self.pasilloAmarillo.append(Casilla(TipoCasilla.pasillo))
				self.pasilloAzul.append(Casilla(TipoCasilla.pasillo))
				self.pasilloRojo.append(Casilla(TipoCasilla.pasillo))
				self.pasilloVerde.append(Casilla(TipoCasilla.pasillo))
			else:
				self.pasilloAmarillo.append(Casilla(TipoCasilla.meta))
				self.pasilloAzul.append(Casilla(TipoCasilla.meta))
				self.pasilloRojo.append(Casilla(TipoCasilla.meta))
				self.pasilloVerde.append(Casilla(TipoCasilla.meta))
	
	def imprimirTablero(self):
		for i in range(len(self.tablero)):
			if len(self.tablero[i].fichas) > 0:
				for ficha in self.tablero[i].fichas:
					print("casilla: ", self.tablero[i].numero, "ficha: ", ficha.color)

File: Backend/parchis-logica/Tablero.js
const Color = require("./Color")
const Casilla = require("./Casilla")
const TipoCasilla = require("./TipoCasilla")

module.exports = class Tablero {
    
    constructor(){
        this.tablero = [];
        this.pasilloAmarillo = []
        this.pasilloAzul = []
        this.pasilloRojo = []
        this.pasilloVerde = []
        this.crearTablero();
    }

    /**
     * Crea el tablero de parchis
     */
    crearTablero(){
        var casillas = [12,17,29,34,46,51,63,68];
        for (let i =1; i<69; i++){
            if(i==5){
                this.tablero.push(new Casilla(TipoCasilla.INICIO,Color.AMARILLO,i))
            }else if(i==22){
                this.tablero.push(new Casilla(TipoCasilla.INICIO,Color.AZUL,i))
            }else if(i==39){
                this.tablero.push(new Casilla(TipoCasilla.INICIO,Color.ROJO,i))
            }else if(i==56){
                this.tablero.push(new Casilla(TipoCasilla.INICIO,Color.VERDE,i))
            }else if(casillas.includes(i)){
                this.tablero.push(new Casilla(TipoCasilla.SEGURA,null,i))
            }else{
                this.tablero.push(new Casilla(TipoCasilla.NORMAL,null,i))
            }
        }
        for (let i =1; i<9; i++){
            this.pasilloAmarillo.push(new Casilla(TipoCasilla.PASILLO,Color.AMARILLO,i));
            this.pasilloAzul.push(new Casilla(TipoCasilla.PASILLO,Color.AZUL,i));
            this.pasilloRojo.push(new Casilla(TipoCasilla.PASILLO,Color.ROJO,i));
            this.pasilloVerde.push(new Casilla(TipoCasilla.PASILLO,Color.VERDE,i));
        }
    }

    /**
     * Devuelve la casilla en la posicion indicada, de forma circular (1 es el primer indice)
     * @param {int} posicion La posicion de la casilla que se quiere obtener.
     * @returns {Casilla} La casilla en la posicion indicada.
     */
    obtenerCasilla(posicion){
        return this.tablero[(posicion-1)%this.tablero.length];
    }

    /**
     * Devuelve el pasillo de un color indicado
     * @param {Color} color El número del color del pasillo que se quiere obtener.
     * @returns Una lista con las casillas del pasillo.
     */
    obtenerListaPasillo(color){
        if(color == Color.AMARILLO){
            return this.pasilloAmarillo;
        }else if(color == Color.AZUL){
            return this.pasilloAzul;
        }else if(color == Color.ROJO){
            return this.pasilloRojo;
        }else if (color == Color.VERDE){
            return this.pasilloVerde;
        }
    }

    /**
     * Obtiene el pasillo de un jugador
     * @param {Jugador} jugador El jugador del que se quiere obtener el pasillo.
     * @param {int} posicion La posicion del pasillo que se quiere obtener.
     * @returns {Casilla} La casilla del pasillo.
     */
    obtenerPasillo(jugador, posicion){
        let lista = this.obtenerListaPasillo(jugador.color);
        if(lista.length < lista.length){
            return lista[posicion];
        }
    }

    /**
     * Crea los pasillos de cada color
     */
    crearPasillos(){
        for(let i=1; i<8; i++){
            if(i<7){
                this.pasilloAmarillo.push(new Casilla(TipoCasilla.PASILLO));
                this.pasilloAzul.push(new Casilla(TipoCasilla.PASILLO));
                this.pasilloRojo.push(new Casilla(TipoCasilla.PASILLO));
                this.pasilloVerde.push(new Casilla(TipoCasilla.PASILLO));
            }else{
                this.pasilloAmarillo.push(new Casilla(TipoCasilla.META));
                this.pasilloAzul.push(new Casilla(TipoCasilla.META));
                this.pasilloRojo.push(new Casilla(TipoCasilla.META));
                this.pasilloVerde.push(new Casilla(TipoCasilla.META));
            }
        }
    }

    /**
     * Imprime el tablero en consola
     */
    imprimirTablero(){
        for(let i=0; i<this.tablero.length; i++){
            if (this.tablero[i].fichas.length > 0){
                this.tablero[i].fichas.forEach(ficha => {
                    console.log("casilla: " + this.tablero[i].numero + " ficha: " + ficha.color);
                });
            }
        }
    }
}